middleware/logging.go

```diff
@@ -89,12 +93,12 @@ func (l Log) Wrap(next http.Handler) http.Handler {
 		if writeErr != nil {
 			if errors.Is(writeErr, context.Canceled) {
 				if l.LogRequestAtInfoLevel {
-					requestLog.Infof("%s %s %s, request cancelled: %s ws: %v; %s", r.Method, uri, time.Since(begin), writeErr, IsWSHandshakeRequest(r), headers)
+					level.Info(requestLog).Log("msg", fmt.Sprintf("%s %s %s, request cancelled: %s ws: %v; %s", r.Method, uri, time.Since(begin), writeErr, IsWSHandshakeRequest(r), headers))
```

Comment:
I'm not a big fan of using `fmt.Sprintf()` here. Also, the main difference is that previously the formatting was deferred until the real logging (which was skipped if the configured log level was higher than the log level) while now we always format it. This is particularly impactful for debug lines. I _think_ you can define a lazy evaluated sprintf, make it implement `Stringer` and pass it here, to get the lazy evaluation (to be tested).

Reply:
Thank you for the hint.